Mark shard failures caused by unsupported aggregations or queries against rolled up data so Kibana can identify them

salvatore-campagna opened the pull request:
When an unsupported aggregation is executed on a
`aggregate_double_metric` field an error with a specific
`type` field is generated. This allows clients like Kibana
to identify these errors and handle them properly.

Also we would like to fail on queries using a date histogram
aggregation with `calendar_interval` on rollup indices.
Date histogram aggregations are executed on rollup indices
only if using `fixed_interval` and a `UTC` timezone.

salvatore-campagna committed:
Handle failures for aggregations on aggregate_double_metric fields

salvatore-campagna committed:
fix: register new elasticsearch exception

salvatore-campagna committed:
fix: register new elasticsearch exception

salvatore-campagna committed:
fix: add the new exception to the ids

salvatore-campagna committed:
fix: prevent possible null pointer exception

salvatore-campagna committed:
fix: correct mistake in null pointer condition check

salvatore-campagna committed:
fix: fail on date histogram aggregations with calendar_interval

To start with we decided that we would like to support only date
histogram aggregations using 'fixed_interval' and throw an error
if 'calendar_interval' is used on a time series index.

salvatore-campagna committed:
fix: use the actual mode for comparison

salvatore-campagna committed:
refactor: improve code readability

salvatore-campagna committed:
fi: fail on date histogram aggregations with non utc timezone

salvatore-campagna committed:
fix: fail on date histogram aggregations with non utc timezone

salvatore-campagna committed:
fix: remove yaml file added by mistake

salvatore-campagna committed:
fix: remove useless min_doc_count

salvatore-campagna committed:
fix: remove empty line added by mistake

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

salvatore-campagna committed:
Merge branch 'main' into feature/unsupported-aggs-downsampling-errors

salvatore-campagna commented on server/src/main/java/org/elasticsearch/ElasticsearchException.java:
```diff
@@ -1578,6 +1578,12 @@ private enum ElasticsearchExceptionHandle {
+            org.elasticsearch.search.aggregations.UnsupportedAggregationOnDownsampledField.class,
+            org.elasticsearch.search.aggregations.UnsupportedAggregationOnDownsampledField::new,
+            167,
+            Version.V_8_4_0
```
Will change this to 8.5.0

salvatore-campagna commented on x-pack/plugin/rollup/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/rollup/20_unsupported_aggs.yml:
```diff
@@ -0,0 +1,164 @@
+setup:
+  - skip:
+      version: " - 8.3.99"
```
Will change this to 8.4.99 and comment with 8.5.0

salvatore-campagna commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregationBuilder.java:
```diff
@@ -417,9 +420,28 @@ protected ValuesSourceAggregatorFactory innerBuild(
 
         final ZoneId tz = timeZone();
+        if (IndexMode.TIME_SERIES.getName().equals(indexMode) && ZoneId.of("UTC").equals(tz) == false) {
+            throw new UnsupportedAggregationOnDownsampledField(
```
Should we use a different exception here?

nik9000 replied:
I'm not sure! I'd go with the same exception for now and see if the kibana folk want to change it.

salvatore-campagna committed:
fix: update version to 8.5.0

salvatore-campagna committed:
fix: add missing skip clause up to version 8.5.0

salvatore-campagna committed:
fix: skip error on null timezone

salvatore-campagna committed:
fix: add missing min_doc_count

salvatore-campagna commented:
@elasticsearchmachine run elasticsearch-ci/part-1

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java:
```diff
@@ -169,6 +171,13 @@ public <T> T getAggregator(RegistryKey<T> registryKey, ValuesSourceConfig values
+                    throw new UnsupportedAggregationOnDownsampledField(
+                        valuesSourceConfig.getDescription() + " is not supported for aggregation [" + registryKey.getName() + "]"
+                    );
+                }
```
I wonder if there's a better place to put this. It feels like kind of a janky spot. @not-napoleon, do you have an opinion on this? Basically we want field types to be able to throw a specific exception if you ask or an unsupported aggregation. Well, really, we just want *this* field to throw it, but I'll bet there is a cleaner way.

salvatore-campagna replied:
Maybe I can do this at the moment the aggregator is created.

salvatore-campagna replied:
This needs to be done for all aggregations. So the only place I see to do it (excluding doing it in every aggregation) is some base class where the construction of the aggregator takes place. Any better idea?

salvatore-campagna replied:
Notice, anyway, that if I remove this piece of code, it fails anyway with 

`throw new IllegalArgumentException(
                    valuesSourceConfig.getDescription() + " is not supported for aggregation [" + registryKey.getName() + "]"
);`

so at the end that is just a way to replace one exception with another.

An option I see is to extract the responsibility of checking for "no value in the registry" with another class that is injected into the ValuesSourceRegistry. A kind of "no such element in the registry" handler.

not-napoleon replied:
I think we already have the handler injected - the `ValuesSourceType` itself can do this.  What you're asking for here is a type-specific behavior.  In aggregations, we don't (generally) work directly with field types, we instead work with the `ValuesSourceType` abstraction, so unless we have a Real Good Reason :TM: that should be the lowest granularity of type switching (real good reasons generally involve "we can optimize this specific index structure, but only if we know about it", for reference).

So, I'd suggest we add a method to the `ValuesSourceType` interface `getUnmappedException(String aggregationName)` or something like that (the VST already has a `typeName()` method it can use for the other string component).  We can add a default implementation that returns the current exception, and override it in `AggergateMetricValuesSourceType.AGGREGATE_METRIC` to throw your new exception.  This also has the advantage of locating the special messaging close to the associated type logic, which seems more natural than hard coding it in the registry.

Let me know if that doesn't work for some reason.

salvatore-campagna replied:
That sound better. Thanks.

salvatore-campagna replied:
I had to add the `ValueSourceConfig` to the parameters because otherwise composing the exact error message was not possible. The `getDescription` has some logic to build it.

nik9000 replied:
I wouldn't bother to check that the exception is non-null here. Returning a `null` supplier isn't really a thing we want to support. I think it's an error is the type doesn't return an exception.

not-napoleon replied:
this needs an else-throw.  If we didn't find a supplier in the registry, we *MUST* throw from this method, or something very weird will break down the road.  Really, it should be impossible for `getUnmappedException` to return `null`, but I'm all for redundant checks.

salvatore-campagna replied:
I see...I thought about that but my idea was "there is the default method". Anyway that is not enough. A subclass overriding the parent method might still return null and then the flow would continue without an exception thrown.

salvatore-campagna replied:
Does it make sense to add a warning message saying something like "returned unexpected null exception, please fix this!"?

nik9000 replied:
You'd asked about double checking that this isn't null rather than just throwing blindly. I'm ok either way. It's a "you built this wrong" sort of error. I don't expect folks to hit it much, but it'd be kind to tell them. I'd use an `assert` for it or not build it at all. But you can do whatever you like here. no big deal to me.

not-napoleon replied:
yeah, thinking about it more, the assert is probably the right choice.  We want to catch this in testing, not give the user an error they can't do anything with, and the assert communicates that more clearly.

salvatore-campagna committed:
doc: include javadoc for the new exception

salvatore-campagna committed:
fix: use the index mode setting directly

salvatore-campagna commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregationBuilder.java:
```diff
@@ -417,9 +420,28 @@ protected ValuesSourceAggregatorFactory innerBuild(
         AggregatorFactories.Builder subFactoriesBuilder
     ) throws IOException {
-        DateHistogramAggregationSupplier aggregatorSupplier = context.getValuesSourceRegistry().getAggregator(REGISTRY_KEY, config);
+        final String indexMode = context.getIndexSettings().getSettings().get(IndexSettings.MODE.getKey());
```
I don't see the method `validateCalendarIntervalType`. Is it something you introduced recently?

salvatore-campagna replied:
Oh...maybe you are suggesting to add it...

salvatore-campagna committed:
refactor: extract a validation method for the calendar interval type

salvatore-campagna committed:
fix: fail validation on 'calendar_interval'

salvatore-campagna committed:
fix: incorrect validation return value

salvatore-campagna committed:
fix: incorrect validation return value

salvatore-campagna committed:
fix: extract logic to throw exceptions from the registry

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregationBuilder.java:
```diff
@@ -417,9 +419,27 @@ protected ValuesSourceAggregatorFactory innerBuild(
-        DateHistogramAggregationSupplier aggregatorSupplier = context.getValuesSourceRegistry().getAggregator(REGISTRY_KEY, config);
+        IndexMode indexMode = context.getIndexSettings().getMode();
+        final DateIntervalWrapper.IntervalTypeEnum dateHistogramIntervalType = dateHistogramInterval.getIntervalType();
+        if (context.getIndexSettings().getMode().validateCalendarIntervalType(dateHistogramIntervalType)) {
```
Can this method be `void` and throw for invalid values? Could you do the same for the time zone below?

salvatore-campagna replied:
Yes I was doing that...returning true/false is confusing...is it returning false if validation is ok or not?

nik9000 replied:
I'd just validate and throw.

salvatore-campagna commented on server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java:
```diff
@@ -92,4 +92,8 @@ default DocValueFormat getFormatter(String format, ZoneId tz) {
      */
     String typeName();
+
+    default RuntimeException getUnmappedException(ValuesSourceConfig config, String name) {
```
Will add the Javadoc.

not-napoleon replied:
I don't love passing a config in here.  I get why, but I'd almost rather we passed in a string and let the caller resolve this.  This is just weirdly recursive, since the config is (generally) the thing that holds the VST.

Maybe this takes two strings, and we add a method to ValuesSourceConfig that calls it with the description? IDK.

Also, I know I suggested the name `getUnmappedException` earlier, but seeing it out of context on `ValuesSourceType` I realized that we use "unmapped" to mean something else generally, and this might be confusing.  How do you feel about `getUnregisteredException` instead?  And either way, this method needs some javadoc.

salvatore-campagna committed:
doc: include javadoc for the new ValuesSourceType interface method

salvatore-campagna committed:
fix: remove file added by mistake

salvatore-campagna committed:
fix: refactor logic to throw exceptions

salvatore-campagna committed:
fix: possible null pointer exception

salvatore-campagna committed:
fix: handle possible null exception providers

salvatore-campagna committed:
fix: void methods do not return any value

salvatore-campagna committed:
fix: assert not null exception

salvatore-campagna committed:
test: assertion error for null values source type exception handler

salvatore-campagna committed:
fix: typo

salvatore-campagna committed:
fix: get rid of unused variable

salvatore-campagna commented:
After discussing with @flash1293 it looks like that the ability to distinguish two types of exceptions would be nice to have. That allows Kibana to have a better handling of errors, telling the user to fix their config if an unsupported aggregation is used on a field, versus telling the user to open a bug because one of the supported aggregations is using an incorrect parameter. This last case applies, at the moment, just to the usage of `calendar_interval` and `non utc timezone` when a date histogram aggregation is executed. For the second type of error the idea is to use an `IllegalArgumentException`.

salvatore-campagna committed:
fix: use illegal_argument_exception for date histogram errors

The date histogram aggregation is one of the aggregations supported
by a time series index. At the moment failures are cause by:
1. using 'calendar_interval' which is not supported
2. using a 'non-utc timezone' which is not supported
For these errors, just using an IllegalArgumentException is enough
and allows clients like Kibana to distinguish agrgegations that are
not supported by agrgegations that are supported but whose query
includes some misconfiguration.

nik9000 reviewed (approved):
LGTM. @not-napoleon, how about you?

nik9000 commented on server/src/main/java/org/elasticsearch/index/IndexMode.java:
```diff
@@ -194,6 +202,25 @@ public DocumentDimensions buildDocumentDimensions() {
         }
+
+        /**
+         * For a time series index we just support 'fixed_interval' aggregations.
```
I think I might have missed something. I thought we only wanted to limit to fixed interval for the rolled up indices. Time series indices are supposed to support all the things. I think. I certainly could be forgetting things, but that was my memory.

flash1293 replied:
good catch, this was also my understanding

salvatore-campagna replied:
What about the UTC/non-UTC check?

flash1293 replied:
Same for UTC/non-UTC by my understanding. Not sure whether it's right or not, but in Kibana we currently check this via the `meta.fixed_interval` property on the field: https://github.com/elastic/kibana/blob/9799dbba27c5baf594357eae0bbfc79b4e7da77c/src/plugins/data_views/server/fetcher/lib/field_capabilities/field_caps_response.ts#L145

csoulios replied:
> I don't get what is wrong here...do you mean this should fail only if the date histogram aggregation runs on a rolled-up index? And not on every time series index? If that is what I missed then I have no idea how do we know if an index is a result of a rollup operation or not. Maybe we have some meta-data? I will check.

It should fail only if the `date_histogram` agg is executed on a rolled up index. "Raw" time-series indices support any `date_histogram` interval. 
Rollup indices contain the `IndexMetadata.INDEX_ROLLUP_SOURCE_NAME` setting in their index settings. And the `@timestamp` field should contain the `fixed_interval` property in the field mapping `meta`. See https://github.com/elastic/elasticsearch/blob/03f3c8119ac1e7bb7034be1f7c4f3355fec4b300/x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java#L440-L453

not-napoleon commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregationBuilder.java:
```diff
@@ -417,9 +417,28 @@ protected ValuesSourceAggregatorFactory innerBuild(
+            .getMode()
+            .validateCalendarIntervalType(
+                dateHistogramIntervalType,
+                config.getDescription()
```
I don't love this pattern of passing the message in and throwing from the validator.  Why not just have the validator return a boolean and throw from here if we want to?  I don't feel terribly strongly about it though.

nik9000 replied:
:rofl: I advocated for having the method do the throwing. I think it's *generally* a cleaner way to do it, especially because the method has to return a very "related" type of exception. So! Maybe it won't matter. This is the thing we're talking about maybe not applying correctly anyway - I think we only want to throw if we're operating on a rolled up index. Not a timeseries one. And *that* will work differently anyway.

salvatore-campagna replied:
I think I did something like this before and I changed it. I find it weird to return a boolean from a validation method.
Should the method return true or false if validation succeeds?

not-napoleon replied:
Maybe I'm splitting hairs, but if the validation method is going to throw, it should own building the message.  That's all I'm saying.  Having the message in one place and the decision to throw in a different place is weird to me.

not-napoleon commented on server/src/test/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistryTests.java:
```diff
@@ -56,4 +58,61 @@ public void testAggregatorNotFoundException() {
         expectThrows(IllegalArgumentException.class, () -> registry.getAggregator(key, scriptOnly));
     }
+
+    public void testValueSourceTypeNullUnregisteredExceptionHandler() {
```
I feel weird saying this, but I think this test is unnecessary to the point of being distracting.  You are literally just testing that the assert works, which isn't going to run in production anyway.  So this test doesn't describe the production behavior, which is a little troubling to me.  I think the assert by itself is enough.

salvatore-campagna replied:
I agree it is pretty much useless by itself, but I wanted to have a test just to stress the point that a non-null exception is required. The other option was to have some Javadoc...but if I have to choose between documentation and a test I usually prefer a test for two reasons:
1. a test fails if something is wrong
2. a test is also a piece of documentation
3. it is very easy to forget updating documentation while you will have a test failing and you have to fix it if something changes

salvatore-campagna replied:
Anyway I can get rid of it.

not-napoleon replied:
Point 2 is exactly my concern - to me, this test documents that if you return a null exception, we still throw, which is what we want.  That seems to say "it's okay to have a null here".

Since our tests run with asserts enabled, a test will fail if we return a null from `getUnregisteredException()`; that's the whole point of the assert being there, and that's the behavior we want to enforce.

salvatore-campagna replied:
Yes right...but if the assert is removed as a result of a code change than something unpredictable can happen. While,  if we have a test checking for the assertion error we would, at least have, a test failing because of the missing assertion. Am I wrong?

salvatore-campagna replied:
Anyway if the assert is removed for whatever reason and `unmappedException` is null then we have:

```
Unexpected exception type, expected AssertionError but got java.lang.NullPointerException: Cannot throw exception because "unmappedException" is null
```

salvatore-campagna replied:
Of course when a production build is created the assertion goes away but at least the CI would have the test checking that the assertion is still there.

not-napoleon reviewed (approved):
Couple of nits, but I think the interaction with the `ValuesSourceRegistry` is okay.

nik9000 reviewed (changes requested):
Hey! officially "requesting changes" while we figure out what's up with that last validation.

salvatore-campagna committed:
test: remove test on assertion

salvatore-campagna committed:
refactor: build exception message inside validation methods

salvatore-campagna committed:
fix: fail date histogram aggregations for rollup indices

Date histogram aggregations must fail if 'calendar_interval' or
a non-utc time zone is used, only if the index the aggregation
is running on is a rollup index. For time series indices, date
histogram aggregations should work as expected without throwing
any error.

salvatore-campagna committed:
Merge branch 'main' into feature/unsupported-aggs-downsampling-errors

salvatore-campagna committed:
fix: code format violations

salvatore-campagna commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -127,6 +128,15 @@ public class IndexMetadata implements Diffable<IndexMetadata>, ToXContentFragmen
         EnumSet.of(ClusterBlockLevel.WRITE)
     );
 
+    public boolean isRollupIndex() {
```
I have the feeling this is not the best I can do here. I am working on a better way to do this at the moment but I wanted to push this change because it might be good enough. If that feeling is shared but this is considered "good enough" we could proceed merging this to unblock Kibana folks experimenting with catching exceptions. Then I can create another issue to refactor this. N.B. I am on vacation next week and Christos is on vacation too and I would like to avoid Kibana folks being stuck waiting for this.

nik9000 replied:
I think it's cool as a temporary thing. Maybe a `TODO` in the code so we know we thought it was temporary when we added it.

flash1293 commented on x-pack/plugin/rollup/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/rollup/20_unsupported_aggs.yml:
```diff
@@ -0,0 +1,164 @@
+  - match: { error.root_cause.0.reason: "Field [total_memory_used] of type [aggregate_metric_double] is not supported for aggregation [percentiles]" }
+
+---
+"Top-metrics aggregation on aggregate_metric_double field":
```
Not sure whether out of scope but what about a test for the mixed case (as this will be extremely common in practice)? Querying both indices at the same time, resulting in one of them succeeding and the other producing a shard failure so there's partial data in the response

salvatore-campagna replied:
Oh that is a good test I think...I am not sure what the result is in this case. I guess we hava a shard failure for the rollup index and partial results coming from the time series index...?

flash1293 replied:
Exactly

salvatore-campagna replied:
This test is not necessary for the errors we throw if the aggregation is not supported (aggregation on field of type `agrgegate_metric_double`). In that case we don't check if the index is a time series index or not...just the field type.

flash1293 reviewed (commented):
Looks mostly good to me, left a nit about another test we could add

salvatore-campagna committed:
test: include a test hitting both indices

Hitting both indices produces partial results coming from the
source time series index and a shard failure coming from the
rollup index.

salvatore-campagna commented on x-pack/plugin/rollup/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/rollup/30_date_histogram.yml:
```diff
@@ -219,3 +219,30 @@ setup:
+                calendar_interval: hour
+                min_doc_count: 1
+
+  - match: { _shards.total: 2 }
```
NOTE: here we rely on the fact that the rollup index is created with the same number of shards of the original index, which is set to 1 in the setup block.

salvatore-campagna committed:
docs: add a note clarifying a test

salvatore-campagna commented on x-pack/plugin/rollup/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/rollup/30_date_histogram.yml:
```diff
@@ -0,0 +1,250 @@
+setup:
```
I had to change this test adding the logic to create the rollup index from the source index because some settings and metadata are applied by the rollup operation and it is not possible to set them from the yaml test. The two settings I use are the source index name and the rollup status which I check to verify if the index is a rollup index or not.

nik9000 commented on server/src/main/java/org/elasticsearch/index/IndexMode.java:
```diff
@@ -196,8 +216,55 @@ public DocumentDimensions buildDocumentDimensions(IndexSettings settings) {
+            String valuesSourceDescription,
+            String aggregationName
+        ) {
+            if (indexSettings.getIndexMetadata().isRollupIndex() && DateIntervalWrapper.IntervalTypeEnum.CALENDAR.equals(intervalType)) {
```
I *think* we want this test for all time rolled up indices. Not that there are any non-time series ones, but I think the addition of `isRollupIndex` means we don't need these new methods in `IndexMode` any more.

nik9000 commented on server/src/main/java/org/elasticsearch/index/IndexMode.java:
```diff
@@ -196,8 +216,55 @@ public DocumentDimensions buildDocumentDimensions(IndexSettings settings) {
+        final boolean rollupSuccess = IndexMetadata.RollupTaskStatus.SUCCESS.name()
+            .toLowerCase(Locale.ROOT)
+            .equals(indexRollupStatus != null ? indexRollupStatus.toLowerCase(Locale.ROOT) : IndexMetadata.RollupTaskStatus.UNKNOWN);
+        return Strings.isNullOrEmpty(sourceIndex) == false && rollupSuccess;
```
I think this is a duplicate.

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/UnsupportedAggregationOnDownsampledField.java:
```diff
@@ -0,0 +1,35 @@
+ * Downsampling uses specific types while aggregating some fields (like 'aggregate_metric_double').
+ * Such field types do not support some aggregations.
+ */
+public class UnsupportedAggregationOnDownsampledField extends AggregationExecutionException {
```
It's probably worth leaving a note that the *name* of this class is part of a contract with Kibana so we shouldn't change it.

nik9000 commented on x-pack/plugin/rollup/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/rollup/20_unsupported_aggs.yml:
```diff
@@ -0,0 +1,164 @@
+                    from: 401.0
+
+  - match: { status: 400 }
+  - match: { error.root_cause.0.type: unsupported_aggregation_on_downsampled_field }
```
It's probably worth leaving a comment here that this `type` name is part of a contract with kibana. Just extra paranoia so folks don't change it.

salvatore-campagna committed:
todo: refactor method isRollupIndex after adding more rollup metadata

salvatore-campagna committed:
fix: remove unused method

salvatore-campagna committed:
fix: todo

salvatore-campagna committed:
note: include a note clarifuing the elasticsearch/kibana contract

salvatore-campagna committed:
refactor: remove date histogram validation methods

csoulios reviewed (commented):
I had a very quick look at the code. I only left a small comment about naming the exception, because this is something we cannot change in the future.

salvatore-campagna committed:
refactor: rename exception

salvatore-campagna committed:
Merge branch 'main' into feature/unsupported-aggs-downsampling-errors

salvatore-campagna committed:
Merge branch 'main' into feature/unsupported-aggs-downsampling-errors

nik9000 commented:
I've merged this so @flash1293 should be able to grab it sooner rather than later.
